app/code/Magento/InventorySales/Model/IsProductSalableForRequestedQtyCondition/IsCorrectQtyCondition.php

```diff
@@ -0,0 +1,104 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\InventorySales\Model\IsProductSalableForRequestedQtyCondition;
+
+use Magento\CatalogInventory\Model\Configuration;
+use Magento\Framework\Math\Division as MathDivision;
+use Magento\InventoryConfigurationApi\Api\GetStockItemConfigurationInterface;
+use Magento\InventoryReservations\Model\GetReservationsQuantityInterface;
+use Magento\InventorySales\Model\GetStockItemDataInterface;
+use Magento\InventorySalesApi\Api\IsProductSalableForRequestedQtyInterface;
+
+/**
+ * @inheritdoc
+ */
+class IsCorrectQtyCondition implements IsProductSalableForRequestedQtyInterface
+{
+    /**
+     * @var GetStockItemConfigurationInterface
+     */
+    private $getStockItemConfiguration;
+
+    /**
+     * @var GetReservationsQuantityInterface
+     */
+    private $getReservationsQuantity;
+
+    /**
+     * @var GetStockItemDataInterface
+     */
+    private $getStockItemData;
+
+    /**
+     * @var Configuration
+     */
+    private $configuration;
+
+    /**
+     * @var MathDivision
+     */
+    protected $mathDivision;
+
+    public function __construct(
+        GetStockItemConfigurationInterface $getStockItemConfiguration,
+        Configuration $configuration,
+        GetReservationsQuantityInterface $getReservationsQuantity,
+        GetStockItemDataInterface $getStockItemData,
+        MathDivision $mathDivision
+    ) {
+        $this->getStockItemConfiguration = $getStockItemConfiguration;
+        $this->configuration = $configuration;
+        $this->getStockItemData = $getStockItemData;
+        $this->getReservationsQuantity = $getReservationsQuantity;
+        $this->mathDivision = $mathDivision;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function execute(string $sku, int $stockId, float $requestedQty): bool
+    {
+        $stockItemConfiguration = $this->getStockItemConfiguration->execute($sku, $stockId);
+        if (null === $stockItemConfiguration) {
+            return false;
+        }
+
+        $stockItemData = $this->getStockItemData->execute($sku, $stockId);
+
+        $qtyWithReservation = $stockItemData['quantity'] + $this->getReservationsQuantity->execute($sku, $stockId);
+        if ($requestedQty > $qtyWithReservation) {
+            return false;
+        }
+
+        $globalMinQty = $this->configuration->getMinQty();
+        if (($stockItemConfiguration->isUseConfigMinQty() == 1 && $requestedQty < $globalMinQty)
+            || ($stockItemConfiguration->isUseConfigMinQty() == 0
+                && $requestedQty < $stockItemConfiguration->getMinQty()
+            )) {
+            return false;
+        }
+
+        $globalMaxSaleQty = $this->configuration->getMaxSaleQty();
```

Comment:
For Quote, we need to preserve behavior from this function 
\Magento\CatalogInventory\Model\StockStateProvider::checkQuoteItemQty

which works with Min Sale Qty 
```

        if ($stockItem->getMinSaleQty() && $qty < $stockItem->getMinSaleQty()) {
            $result->setHasError(true)
                ->setMessage(__('The fewest you may purchase is %1.', $stockItem->getMinSaleQty() * 1))
                ->setErrorCode('qty_min')
                ->setQuoteMessage(__('Please correct the quantity for some products.'))
                ->setQuoteMessageIndex('qty');
            return $result;
        }

        if ($stockItem->getMaxSaleQty() && $qty > $stockItem->getMaxSaleQty()) {
            $result->setHasError(true)
                ->setMessage(__('The most you may purchase is %1.', $stockItem->getMaxSaleQty() * 1))
                ->setErrorCode('qty_max')
                ->setQuoteMessage(__('Please correct the quantity for some products.'))
                ->setQuoteMessageIndex('qty');
            return $result;
        }
```

That actually means that we need to have 2 chains:
- one for Quote
- another for Checkout